File: binary_search.py
import math

def binary_search(haystack: list[int], number: int) -> bool:
    """ NOTE: Only works on a sorted array """
    low = 0
    high = len(haystack)

    while low < high:
        mid = math.floor( low + ((high-low) /2 ))
        value = haystack[mid]

        if value == number:
            return True
        elif value > number:
            high = mid
        elif value < number:
            low = mid + 1

    return False
        

if __name__ == "__main__":
    haystack: list[int] = [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14 , 15, 16, 17]
    result = binary_search(haystack, 7)
    print(result)

File: src/bin/binary_search.rs
/// NOTE: Primary conditions for binary search
/// - Value equals target
/// - Value is larger than target
/// - Value is smaller than target
fn binary_search(haystack: [i16; 9], needle: i16) -> bool {
    let mut low = 0;
    let mut high = haystack.len();

    while low < high {
        let mid = low + ((high - low) / 2);
        let value = haystack[mid as usize];

        if value == needle {
            return true;
        } else if value > needle {
            high = mid
        } else {
            low = mid + 1
        }
    }

    return false;
}

fn main() {
    let haystack = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    let needle = 5;
    let found = binary_search(haystack, needle);
    println!("{}", found);
}

#[cfg(test)]
mod tests {
    use super::binary_search;

    #[test]
    fn test_found() {
        let haystack = [1, 2, 3, 4, 5, 6, 7, 8, 9];
        let needle = 5;

        assert_eq!(binary_search(haystack, needle), true)
    }

    #[test]
    fn test_not_found() {
        let haystack = [1, 2, 3, 4, 6, 7, 8, 9, 11];
        let needle = 5;

        assert_eq!(binary_search(haystack, needle), false)
    }
}
